README.md
# Sequence ID Generator

A small library for generating base62 encoded sequence. It can be used to generate incremental sequence but also to generate random values.

Random values are generated with a Linear congruential generator - each value in the sequence is generated only once. After all possible values are generated, the sequence is repeated.

## Usage

### Default behavior

By default, base62 encoded values from an interval <0 Number.MAX_SAFE_INTEGER) are generated.

```js
import { createGenerator } from 'sequence-id-generator';

const rnd = createGenerator();
for (let i = 0; i < 15; i++) {
    console.log(rnd());
}
// output: 0,1,2,3,4,5,6,7,8,9,a,b,c,d,e
```

### Custom alphabet

```js
const rnd = createGenerator({
    alphabet: '01234',
});
for (let i = 0; i < 10; i++) {
    console.log(rnd());
}
// output: 0,1,2,3,4,10,11,12,13,14
```

### Limit generated values

Setting min and max value.

```js
const rnd = createGenerator({
    alphabet: '01234',
    max: 4,
});
for (let i = 0; i < 10; i++) {
    console.log(rnd());
}
// output: 0,1,2,3,4,0,1,2,3,4
```

```js
const rnd = createGenerator({
    alphabet: '01234',
    min: 1
    max: 8,
});
for (let i = 0; i < 10; i++) {
    console.log(rnd());
}
// output: 1,2,3,4,10,11,12,13,1,2
```

Setting minLength and maxLength will calculate and override min and max options.

```js
const rnd = createGenerator({
    alphabet: '012',
    minLength: 2,
    maxLength: 3,
});
for (let i = 0; i < 10; i++) {
    console.log(rnd());
}
// output: 10,11,12,20,21,22,100,101,102,110
```

### Random values

Shuffle alphabet - value will be increased by 1 but output will be encoded with a shuffled alphabet.

```js
const rnd = createGenerator({
    alphabet: '01234',
    shuffle: true,
});
for (let i = 0; i < 10; i++) {
    console.log(rnd());
}
// output: 3,1,4,0,2,13,11,14,10,12
```

Random generator - each value from sequence is generated exactly once.

```js
const rnd = createGenerator({
    alphabet: '01234',
    max: 8,
    random: true,
});
for (let i = 0; i < 20; i++) {
    console.log(rnd());
}
// output: 12,2,11,1,10,0,4,13,3,12,2,11,1,10,0,4,13,3,12,2
```

### Fixed string length

```js
const rnd = createGenerator({
    alphabet: '01234',
    maxLength: 3,
    random: true,
    fixedLength: true,
});
for (let i = 0; i < 20; i++) {
    console.log(rnd());
}
// output: 102,200,223,031,414,132,030,403,011,144
```

## Options

| Name        | Description                                                                                        | Default                                                         |
| ----------- | -------------------------------------------------------------------------------------------------- | --------------------------------------------------------------- |
| min         | Inclusive minimum generated value. 0 <= min < Number.MAX_SAFE_INTEGER                              | 0                                                               |
| max         | Exclusive maximum generated value. min < max <= Number.MAX_SAFE_INTEGER                            | Number.MAX_SAFE_INTEGER                                         |
| minLength   | Minimal length of generated encoded string. If specified, min option is ignored.                   | 0                                                               |
| maxLength   | Maximal length of generated encoded string. If specified, max option is ignored.                   | 9                                                               |
| fixedLength | If enabled, all generated strings has the same length = maxLength.                                 | false                                                           |
| prefix      | Prefix used if fixedLength is enabled. Default value us first character from (shuffled) alphabet.  | '0'                                                             |
| shuffle     | If enabled, alphabet is shuffled.                                                                  | false                                                           |
| random      | If enabled, random values within range <min, max) will be generated instead of incremental values. | false                                                           |
| alphabet    | Custom alphabet used for encoding.                                                                 | '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ |

## API

### Object properties

-   options: get all used options including calculated ones.

```js
const rnd = createGenerator();
console.log(rnd.options);
// output:
// {
//     min: 0,
//     max: 9007199254740991,
//     minLength: 1,
//     maxLength: 9,
//     prefix: '0',
//     fixedLength: false,
//     random: false,
//     shuffle: false,
//     alphabet: '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ'
// }
```


CHANGELOG.md
# Change Log

## 0.1.0 (2023-01-04)

-   Added esm build.
-   Changed export from default to named.
    Use _import { createGenerator } from 'sequence-id-generator'_ instead of _import createGenerator from 'sequence-id-generator'_.

## 0.0.2 (2023-01-03)

-   Added validation
-   Added readme

## 0.0.1 (2023-01-01)

-   First release


src/lcg.test.ts
import { createLcg } from './lcg';

describe('lcg', () => {
    test('generates full range from 0', () => {
        const range = 235236;
        const rnd = createLcg(0, range);
        const generated: number[] = [];
        const indexes: number[] = [];

        for (let index = 0; index < range; index++) {
            indexes.push(index);
            generated.push(rnd());
        }
        generated.sort((a, b) => a - b);
        expect(generated).toEqual(indexes);
    });

    test('generates full range', () => {
        const rnd = createLcg(10, 20);

        const generated: number[] = [];
        const indexes: number[] = [];

        for (let index = 10; index < 20; index++) {
            indexes.push(index);
            generated.push(rnd());
        }
        generated.sort((a, b) => a - b);
        expect(generated).toEqual(indexes);
    });

    test('generates same cycle', () => {
        const range = 235236;
        const rnd = createLcg(0, range);
        const firstCycle: number[] = [];
        const secondCycle: number[] = [];

        for (let index = 0; index < range; index++) {
            firstCycle.push(rnd());
        }
        for (let index = 0; index < range; index++) {
            secondCycle.push(rnd());
        }
        expect(firstCycle).toEqual(secondCycle);
    });

    test('multiplier is correct', () => {
        expect(createLcg(0, 853).multiplier).toEqual(854);
        expect(createLcg(0, 1000).multiplier).toEqual(41);
        expect(createLcg(0, 1000000).multiplier).toEqual(21);
        expect(createLcg(0, 235236).multiplier).toEqual(235237);
    });

    test('increment is correct', () => {
        expect(createLcg(0, 13 * 17).increment).toEqual(19);
        expect(createLcg(0, 5).increment).toEqual(13);
        expect(createLcg(0, 1000).increment).toEqual(13);
    });
});


src/index.test.ts
import { createGenerator, Generator } from './index';

describe('generator', () => {
    test('default options are correctly created', () => {
        const gen = createGenerator();

        expect(gen.options).toEqual({
            alphabet: '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ',
            fixedLength: false,
            max: 9007199254740991,
            maxLength: 9,
            min: 0,
            minLength: 1,
            prefix: '0',
            random: false,
            shuffle: false,
        });
    });

    test('options - shuffled alphabet', () => {
        const gen = createGenerator({
            alphabet: '01234',
            shuffle: true,
        });
        expect(gen.options.alphabet).not.toBe('01234');
        expect(gen.options.alphabet.split('').sort().join('')).toBe('01234');
    });

    test('options - correct minLength and maxLength calculated', () => {
        let gen = createGenerator({
            alphabet: '01234',
            min: 0,
            max: 5,
        });
        expect(gen.options.minLength).toBe(1);
        expect(gen.options.maxLength).toBe(2);

        gen = createGenerator({
            alphabet: '01234',
            min: 4,
            max: 6,
        });
        expect(gen.options.minLength).toBe(1);
        expect(gen.options.maxLength).toBe(2);

        gen = createGenerator({
            alphabet: '01234',
            min: 24,
            max: 25,
        });
        expect(gen.options.minLength).toBe(2);
        expect(gen.options.maxLength).toBe(3);
    });

    test('options - correct min and max calculated', () => {
        let gen = createGenerator({
            alphabet: '01234',
            minLength: 1,
            maxLength: 1,
        });
        expect(gen.options.min).toBe(0);
        expect(gen.options.max).toBe(4);

        gen = createGenerator({
            alphabet: '01234',
            minLength: 2,
            maxLength: 2,
        });
        expect(gen.options.min).toBe(5);
        expect(gen.options.max).toBe(24);

        gen = createGenerator({
            alphabet: '01234',
            min: 0,
            max: 5,
            minLength: 2,
            maxLength: 3,
        });
        expect(gen.options.min).toBe(5);
        expect(gen.options.max).toBe(124);

        gen = createGenerator({
            minLength: 25,
            maxLength: 25,
        });
        expect(gen.options.min).toBe(Number.MAX_SAFE_INTEGER - 1);
        expect(gen.options.max).toBe(Number.MAX_SAFE_INTEGER);
    });

    test('generate values', () => {
        let gen: Generator;
        let values: string[];

        gen = createGenerator({
            alphabet: '01234',
        });
        values = [];
        for (let i = 0; i <= 15; i++) {
            values.push(gen());
        }
        expect(values).toEqual([
            '0',
            '1',
            '2',
            '3',
            '4',
            '10',
            '11',
            '12',
            '13',
            '14',
            '20',
            '21',
            '22',
            '23',
            '24',
            '30',
        ]);

        gen = createGenerator({
            alphabet: '01234',
            max: 5,
        });
        values = [];
        for (let i = 0; i <= 11; i++) {
            values.push(gen());
        }
        expect(values).toEqual(['0', '1', '2', '3', '4', '10', '0', '1', '2', '3', '4', '10']);
    });

    test('generate shuffled values', () => {
        let gen: Generator;
        let values: string[];

        gen = createGenerator({
            alphabet: '01234',
            shuffle: true,
            max: 4,
        });
        values = [];
        for (let i = 0; i < 5; i++) {
            values.push(gen());
        }
        expect(values).not.toEqual(['0', '1', '2', '3', '4']);
        expect(values.sort()).toEqual(['0', '1', '2', '3', '4']);
    });

    test('generate random values', () => {
        let gen: Generator;
        let values: string[];

        gen = createGenerator({
            alphabet: '01234',
            random: true,
            max: 10,
        });
        values = [];
        for (let i = 0; i < 12; i++) {
            values.push(gen());
        }
        expect(values).not.toEqual(['0', '1', '2', '3', '4', '10', '11', '12', '13', '14', '20', '0']);
        expect(values[0]).toBe(values[values.length - 1]);
    });

    test('generate values with prefix', () => {
        let gen: Generator;
        let values: string[];

        gen = createGenerator({
            alphabet: '01234',
            max: 10,
            fixedLength: true,
        });
        values = [];
        for (let i = 0; i < 12; i++) {
            values.push(gen());
        }
        expect(values).toEqual(['00', '01', '02', '03', '04', '10', '11', '12', '13', '14', '20', '00']);

        gen = createGenerator({
            alphabet: '01234',
            max: 10,
            fixedLength: true,
            prefix: '-',
        });
        values = [];
        for (let i = 0; i < 12; i++) {
            values.push(gen());
        }
        expect(values).toEqual(['-0', '-1', '-2', '-3', '-4', '10', '11', '12', '13', '14', '20', '-0']);
    });
});


src/index.ts
import { createLcg } from './lcg';
import { base62, shuffleArray } from './utils';

export interface GeneratorOptions {
    /** Min generated value (inclusive). */
    min: number;
    /** Max generated value (inclusive). */
    max: number;
    /** Min length of generated string. */
    minLength: number;
    /** Max length of generated string. */
    maxLength: number;
    /** All generated strings will be prefixed and have maxLength length. */
    fixedLength: boolean;
    /** Prefix used if fixedLength in true. If not specified, first alphabet letter is used. */
    prefix: string;
    /** If true, alphabet will be shuffled. */
    shuffle: boolean;
    /**
     * If true, random values will be generated.
     * Each number between min and max will be generated exactly once.
     * After all possible values are generated, it repeats the same random sequence again.
     */
    random: boolean;
    /**
     * Custom alphabet.
     * Default is "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
     */
    alphabet: string;
}

export interface Generator {
    (options?: GeneratorOptions): string;
    readonly options: GeneratorOptions;
}

const DEFAULT_ALPHABET = '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ';

export function createGenerator(opts?: Partial<GeneratorOptions>): Generator {
    const options: GeneratorOptions = getOptions(opts);
    const alphabet: string[] = options.alphabet.split('');

    const gen = getNumberGenerator(options.random, options.min, options.max);

    function generate(): string {
        const minChars = options.fixedLength ? options.maxLength : options.minLength;
        return base62(gen(), alphabet, options.prefix, minChars);
    }

    return Object.assign(generate, {
        get options(): GeneratorOptions {
            return { ...options };
        },
    });
}

function getNumberGenerator(random: boolean, min: number, max: number) {
    if (random) {
        return createLcg(min, max + 1);
    }

    let index: number = min;
    return function () {
        const i = index;
        index++;
        if (index > max) {
            index = min;
        }
        return i;
    };
}

function getOptions(options?: Partial<GeneratorOptions>): GeneratorOptions {
    const alphabet = getAlphabet(options?.alphabet ?? DEFAULT_ALPHABET, options?.shuffle);
    if (!alphabet.length) {
        throw new Error('Options Error: alphabet cannot be empty.');
    }

    let min: number;
    let max: number;
    let minLength: number;
    let maxLength: number;

    if (options?.minLength !== undefined) {
        if (options.minLength < 1) {
            throw new Error('Options Error: minLength has to be at least 1.');
        }
        minLength = options.minLength;
        min = decodedMinLength(options.minLength, alphabet.length);
    } else {
        min = options?.min ?? 0;
        minLength = encodedLength(min, alphabet.length);
    }
    if (options?.maxLength !== undefined) {
        if (options.maxLength < 1) {
            throw new Error('Options Error: maxLength has to be at least 1.');
        }
        maxLength = options.maxLength;
        max = decodedMaxLength(options.maxLength, alphabet.length);
    } else {
        max = options?.max ?? Number.MAX_SAFE_INTEGER;
        maxLength = encodedLength(max, alphabet.length);
    }

    if (min < 0) {
        throw new Error('Options Error: min has to be at least 0.');
    }
    if (!Number.isSafeInteger(min) || !Number.isSafeInteger(max)) {
        throw new Error('Options Error: min and max have to be safe integers.');
    }
    if (min >= max) {
        throw new Error('Options Error: min has to be lower than max.');
    }

    return {
        min,
        max,
        minLength,
        maxLength,
        prefix: options?.prefix || alphabet.charAt(0),
        fixedLength: options?.fixedLength ?? false,
        random: options?.random ?? false,
        shuffle: options?.shuffle ?? false,
        alphabet,
    };
}

function decodedMinLength(n: number, alphabetLength: number): number {
    if (n === 1) {
        return 0;
    }
    let min = 1;
    for (let i = 0; i < n - 1; i++) {
        min = min * alphabetLength;
        if (!Number.isSafeInteger(min) || min < 0) {
            return Number.MAX_SAFE_INTEGER - 1;
        }
    }
    return min;
}

function decodedMaxLength(n: number, alphabetLength: number): number {
    let max = 1;
    for (let i = 0; i < n; i++) {
        max = max * alphabetLength;
        if (!Number.isSafeInteger(max) || max < 0) {
            return Number.MAX_SAFE_INTEGER;
        }
    }
    return max - 1;
}

function encodedLength(n: number, alphabetLength: number) {
    if (n === 0) {
        return 1;
    }
    let l = 0;
    let b = 1;
    while (b <= n) {
        l++;
        b = b * alphabetLength;
    }
    return l;
}

function getAlphabet(alphabet: string, shuffle: boolean | undefined): string {
    const chars = alphabet.split('');
    if (shuffle) {
        shuffleArray(chars);
    }
    return chars.join('');
}


src/utils.ts
export function shuffleArray(array: string[]) {
    let currentIndex = array.length;
    let randomIndex: number;

    while (currentIndex != 0) {
        randomIndex = Math.floor(Math.random() * currentIndex);
        currentIndex--;
        [array[currentIndex], array[randomIndex]] = [array[randomIndex], array[currentIndex]];
    }
}

export function base62(i: number, alphabet: string[], prefix?: string, minChars?: number): string {
    let c: string;
    if (i === 0) {
        c = alphabet[0];
    } else {
        let sb = '';
        while (i > 0) {
            const rem = i % alphabet.length;
            sb = alphabet[rem] + sb;
            i = Math.floor(i / alphabet.length);
        }

        c = sb;
    }

    if (prefix && minChars) {
        return c.padStart(minChars, prefix);
    } else {
        return c;
    }
}


src/Generator.ts
import { shuffle } from 'lodash';

const BASE_ALPHABET = '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ';
const ALPHABET_LENGTH = BASE_ALPHABET.length;
const MAX_LENGTH = 8;

function generateAlphabet(random: boolean): string[] {
    const chars = BASE_ALPHABET.split('');
    if (random) {
        return shuffle(chars);
    }
    return chars;
}

export interface GeneratorOptions {
    minChars?: number;
    maxChars?: number;
    prefix?: string;
    shuffle?: boolean;
}

export class Generator {

    private readonly options: GeneratorOptions;
    private readonly alphabet: string[];
    private readonly minIndex: number;
    private readonly maxIndex: number;
    private index: number;

    constructor(options?: GeneratorOptions) {
        this.options = Object.assign({
            minChars: 1,
            maxChars: MAX_LENGTH,
            shuffle: true,
        }, options);

        if (!this.options.minChars || this.options.minChars < 1 || this.options.minChars > MAX_LENGTH) {
            throw new Error(`minChars has to be between 1 and ${MAX_LENGTH}`);
        }
        if (!this.options.maxChars || this.options.maxChars < 1 || this.options.maxChars > MAX_LENGTH) {
            throw new Error(`maxChars has to be between 1 and ${MAX_LENGTH}`);
        }

        this.alphabet = generateAlphabet(!!this.options.shuffle);
        this.minIndex = (this.options.minChars === 1) ? 0 : Math.pow(ALPHABET_LENGTH, this.options.minChars - 1);
        this.maxIndex = Math.pow(ALPHABET_LENGTH, this.options.maxChars) - 1;
        this.index = this.minIndex;
    }

    get sequenceLength(): number {
        return this.maxIndex - this.minIndex + 1;
    }

    generate(): string {
        const id = this.encodeIndex(this.index);
        this.index++;
        if (this.index > this.maxIndex) {
            this.index = this.minIndex;
        }

        return id;
    }

    private encodeIndex(index: number): string {
        return this.base62(index, this.alphabet);
    }

    private base62(i: number, alphabet: string[]): string {
        let c: string;
        if (i === 0) {
            c = alphabet[0];
        } else {
            let sb = '';
            while (i > 0) {
                const rem = i % ALPHABET_LENGTH;
                sb = alphabet[rem] + sb;
                i = Math.floor(i / ALPHABET_LENGTH);
            }

            c = sb;
        }

        if (this.options?.prefix) {
            return c.padStart(this.options.maxChars || MAX_LENGTH, this.options.prefix);
        } else {
            return c;
        }
    }
}


src/lcg.ts
const primes = [13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97];

export interface LcgGenerator {
    (): number;
    readonly increment: number;
    readonly modulus: number;
    readonly multiplier: number;
}

/**
 * @param min min generated value (inclusive)
 * @param max max generated value (exclusive)
 * @param seed custom seed (for testing)
 */
export function createLcg(min: number, max: number, seed?: number): LcgGenerator {
    if (!Number.isInteger(min) || !Number.isInteger(max)) {
        throw new Error('min and max have to be integers.');
    }
    if (min < 0) {
        throw new Error('min has to be >= 0.');
    }
    if (!Number.isSafeInteger(max)) {
        throw new Error(`max has to be <= ${Number.MAX_SAFE_INTEGER}.`);
    }

    const modulus = max - min;
    const increment = getIncrement(modulus);
    const multiplier = getMultiplier(modulus);
    let value: number;
    if (typeof seed === 'number') {
        if (seed < modulus) {
            value = seed;
        } else {
            value = seed % modulus;
        }
    } else {
        value = Math.floor(Math.random() * Number.MAX_SAFE_INTEGER) % modulus;
    }

    function generator() {
        value = (multiplier * value + increment) % modulus;
        return value + min;
    }

    return Object.assign(generator, {
        get increment() {
            return increment;
        },
        get modulus() {
            return modulus;
        },
        get multiplier() {
            return multiplier;
        },
    });
}

function getIncrement(num: number): number {
    for (const prime of primes) {
        if (num % prime !== 0) {
            return prime;
        }
    }
    return primes[0];
}

function primeFactors(num: number): number[] {
    const factors: number[] = [];
    let sqrt = Math.sqrt(num);
    for (let i = 3; i < sqrt; i++) {
        if (num % i === 0) {
            do {
                num = Math.floor(num / i);
            } while (num % i === 0);

            sqrt = Math.sqrt(num);
            factors.push(i);
        }
    }

    if (num > 1) {
        factors.push(num);
    }

    return factors;
}

function getMultiplier(num: number): number {
    let multiplier = 1;

    for (const factor of primeFactors(num)) {
        multiplier *= factor;
    }
    while (multiplier % 4 !== 0 && num % 4 === 0) {
        multiplier *= 2;
    }

    return multiplier + 1;
}
